src/App.php

```diff
@@ -657,37 +657,49 @@ public function stickyForget(string $name): void
      */
     public function url($page = [], array $extraRequestUrlArgs = []): string
     {
-        $request = $this->getRequest();
-
-        $pagePath = '';
         if (is_string($page)) {
             $pageExploded = explode('?', $page, 2);
-            $pagePath = $pageExploded[0];
             parse_str($pageExploded[1] ?? '', $page);
+            $pagePath = $pageExploded[0] !== '' ? $pageExploded[0] : null;
         } else {
-            if (isset($page[0])) {
-                $pagePath = $page[0];
-            } else {
-                // use current page by default
-                $requestUrl = $request->getUri()->getPath();
-                if (substr($requestUrl, -1, 1) === '/') {
-                    $pagePath = $this->urlBuildingIndexPage;
-                } else {
-                    $pagePath = basename($requestUrl, $this->urlBuildingExt);
-                }
-            }
+            $pagePath = $page[0] ?? null;
             unset($page[0]);
-            if ($pagePath) {
-                $pagePath .= $this->urlBuildingExt;
+        }
+
+        $request = $this->getRequest();
+
+        if ($pagePath === null) {
+            $pagePath = $request->getUri()->getPath();
+        }
+
+        $prefix = '';
+        if (substr($pagePath, 0, 2) === '..') {
+            $prefix = '..';
```

Comment:
this is untested and limited to one level only, if "dots" are supported, then "." must be supported as well

@abbadon1334 did you coded this by yourself or used a GPT to code it?

it took me 2 hours to clean your code and the tests...

Reply:
The code started as clear but added some conditional to solve some issues in demos, e.g. the `../` and the ./ which are the same as empty string due to browser behavior.
sorry @mvorisek, but due to the mess and tiredness, i run at the end rector for renaming variables and method names, it seems more clear and i left it, next time i commit the actual version and do a rector just to go back if there is any problem

Reply:
most non-BC-breaking changes were fixed/landed in #2108